### PR TITLE
[MM-35613] Self-managed hosting, with AWS Lambda support

### PR DESCRIPTION
**Pull request by levb**
This is the next iteration on https://github.com/mattermost/mattermost-plugin-apps/pull/181 (will close soon).

Due to number of changes I decided that a new clean PR is a better course of action, even for those who have already reviewed the original (sans AWS).

### DEMO/Test:
1. Set the following environment variables, use a personal AWS account, substitute the bucket/creds as appropriate:
```
export MM_APPS_AWS_REGION=us-west-2
export MM_APPS_S3_BUCKET=My-existing-bucket
export MM_APPS_PROVISION_AWS_ACCESS_KEY=****E7PF
export MM_APPS_PROVISION_AWS_SECRET_KEY=****TOKX
```

2. Initialize the AWS resources:
```
go run ./cmd/appsctl aws init --create --create-access-key
```
The output of the command will contain the "Invoke" access key, make sure you add it to the environment.

**Restart your Mattermost server with the new variables set**

3. Build `hello-lambda` bundle
```
cd ./examples/go/hello-lambda && make dist
```

4. Test the environment
```
go run ./cmd/appsctl aws test provision ./examples/go/hello-lambda/dist/bundle.zip
go run ./cmd/appsctl aws test lambda
go run ./cmd/appsctl aws test s3
```
(if the changes do not appear to work, resulting in 403/404 errors - wait for several minutes then refresh again)

5. Re-provision the `hello-lambda` again, to see the output
```
go run ./cmd/appsctl provision app ./examples/go/hello-lambda/dist/bundle.zip --update
```
Note that you should end up with the following _invoke_ policy. As more Apps are added, their functions will be added to the `AllowLambda` statement.
image

6. Install the `hello-lambda` with the command displayed in the prior step. (REFRESH if the changes do not appear to work, resulting in 403/404 errors - wait for several minutes then refresh again)

### Notes
1. new: **Cloud** vs **Self-managed** mode, cloud mode is on if the License is Cloud, or in case of `community.mattermost.com` is triggered on by the presence of otherwise deprecated `APPS_INVOKE_AWS_ACCESS_KEY` environment variable.
1. removed: `appsctl provision bucket`, the bucket must be created and provided by the customer.
2. new: `appsctl aws init [options...]`, initialize self-managed AWS environment to provision and use Apps.
3. new: `appsctl aws clean`, (best attempt, unfinished) to cleanup default entities created for Apps.
4. new: `appsctl aws test`, provision `hello-lambda` example App and test access to its resources (can be further improved).
5. new: environment variables to document, see https://github.com/mattermost/mattermost-plugin-apps/blob/74ef34fa8aa98309031ab78e2513f81f585b7e21/upstream/upaws/aws.go#L9-L23.
6. new: `/apps install aws $APP_ID $VERSION` in self-managed mode.e.g. `/apps install hello-lambda demo`
7. After an apps is uninstalled in on-prem, the manifest is still listed.

Other
- Marketplace app manifests are only initialized in cloud mode
- Combined all AWS access code (Client, Provisioner, Upstream) in `upstream/upaws`
- Refactored command initialization

#### Install commands

- `/apps install marketplace com.mattermost.servicenow` is used in Cloud to install from the Marketplace
- `/apps install http http://localhost:3000/manifest --app-secret abc` is used in on-prem to install from an HTTP server
- `/apps install aws com.mattermost.servicenow 0.1.0` is used in on-prem to install an App that was previously deployed to cloud

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-35613

### TODO
- Add `hello-lambda` README.md
- Documentation

### CODE REVIEW

File: apps/manifest.go
Comment (levb) [inline apps/manifest.go:104]: 0/5 Makes it easier to have it here rather than try to abstract type-specific structs; may still come as a separate PR. Other AWS-related types etc are now in `upstrem/upaws`

File: e2e/cypress/integration/bindings/channel_header_spec.ts
Comment (levb) [inline e2e/cypress/integration/bindings/channel_header_spec.ts:17]: The E2E test now run in the "self-managed" mode, so install by URL works.

File: examples/go/hello-lambda/Makefile
Comment (levb) [inline examples/go/hello-lambda/Makefile:1]: Need README.MD

File: examples/go/hello-lambda/pong.json
Comment (levb) [inline examples/go/hello-lambda/pong.json:1]: Used by `appsctl aws test lambda`

File: examples/go/hello-lambda/static/test.txt
Comment (levb) [inline examples/go/hello-lambda/static/test.txt:1]: used by `appsctl aws test s3`

File: upstream/upaws/aws.go
Comment (levb) [inline upstream/upaws/aws.go:1]: @stylianosrigas this one is relevant for you to look at

File: upstream/upaws/initialize.go
Comment (levb) [inline upstream/upaws/initialize.go:35]: @stylianosrigas also here - the initialization of policies, roles, user, group.

**Comment by DHaussermann** [comment]
@levb when testing in Cloud mode, the instructions above work for the most part. I saw couple issues with this...
1. It seems as though the stock permissions applied when creating a bucket do not work. When creating a bucket the default setting is block all public access. When left unchanged I saw the follwoing:
image
Making the following permission change resolved the problem I was seeing
image
Please advise if this is expected or acceptable.
2. As per point 6. above, new variable names should be explicitly documented. Seeing `"invoke"` in step 2. confused me as it seemed the variable names had not changed. (The new banes are in the terminal window but I still made a user error here)
3. The instruction set should explicitly state which environment variables are needed by the appsctl tool and which are needed by the server. I did this without updating the `bash` and the install process was failing due to the server missing some variables. 

Regression testing of Cloud mode is still ongoing. Also have not attempted to use the AWS clean commands.

File: cmd/appsctl/main.go
Comment (hanzei) [inline cmd/appsctl/main.go:44]: What is the difference to `AsTest`?

File: examples/go/hello-lambda/Makefile
Comment (hanzei) [reply examples/go/hello-lambda/Makefile:1]: Separate ticket?

File: examples/go/hello-lambda/bindings.json
Comment (hanzei) [inline examples/go/hello-lambda/bindings.json:11]:
```suggestion
					"hint": "[send]",
```

File: examples/go/hello-lambda/main.go
Comment (hanzei) [inline examples/go/hello-lambda/main.go:44]: 1/5 that it's idiomatic to define the manifest as a go variable and rewrite based on the mode to keep things DRY.

File: examples/go/hello-lambda/manifest-http.json
Comment (hanzei) [inline examples/go/hello-lambda/manifest-http.json:3]: Nit pick: Version is missing

File: examples/go/hello-lambda/Makefile
Comment (hanzei) [inline examples/go/hello-lambda/Makefile:1]: Should we also rename and update `js/aws_hello`? Separate PR /Ticket?

File: go.mod
Comment (hanzei) [inline go.mod:8]: 2/5 that this should go into a seperate `go.mod` file just for the example to make it easier to copy and run

File: server/command/service.go
Comment (hanzei) [inline server/command/service.go:33]: 0/5 `commandParams`?

File: server/command/service.go
Comment (hanzei) [inline server/command/service.go:58]: Should the visibility be limited to system admins?

File: server/command/service.go
Comment (hanzei) [inline server/command/service.go:126]: Is that option really needed for Cloud?

File: server/command/service.go
Comment (hanzei) [inline server/command/service.go:82]: Given that is coding is getting touched: Should `s.checkSystemAdmin` be added here for consistency?

File: server/plugin.go
Comment (hanzei) [inline server/plugin.go:101]: Does it make sense to also do that in dev mode given that the install command for AWS is available?

File: server/proxy/proxy.go
Comment (hanzei) [inline server/proxy/proxy.go:204]: Similar HTTP apps should only run in On-Prem. Should there be a check to enforce that?

File: upstream/upaws/clean.go
Comment (hanzei) [inline upstream/upaws/clean.go:27]: Why not
```suggestion
	err := asAdmin.RemoveUserFromGroup(DefaultUserName, DefaultGroupName)
```

File: upstream/upaws/clean.go
Comment (hanzei) [inline upstream/upaws/clean.go:13]: Nice anonymous function :+1:

File: upstream/upaws/clean.go
Comment (hanzei) [inline upstream/upaws/clean.go:77]: Ticket?

**Review by hanzei — CHANGES_REQUESTED** [review]
Great work :+1: 

The code looks solid to me, but I'm running into the same issue as @DHaussermann in 1.

**Comment by levb** [comment]
@hanzei @DHaussermann opening up the bucket to ACLs works for now, 
image

I propose that instead of making the manifest file public (which is causing the exception) we can just proxy the entire /manifests folder from the MM instance, making all manifests there visible. Separate PR?

**Comment by hanzei** [comment]
@levb I'm struggling to understand what does changes with your PR. The ACLs weren't public before and should not have to, right?

**Comment by levb** [comment]
@hanzei the provisioning of static assets shouldn't have changed at all, and I am not seeing them work from the server when the bucket is private. I question if they ever worked (we don't use provisioned icons in the apps at the moment)

**Review by hanzei — APPROVED** [review]
Tested again and works now :tada: 

Submitted by feedback in https://github.com/mattermost/mattermost-plugin-apps/pull/193 for quicker iteration

LGTM :+1:

**Comment by DHaussermann** [comment]
@levb one issue that I see with this is that the `appsctl aws init` command cannot be run a 2nd time with a different bucket. If I set a different bucket and run the script again, as expected, I see `INFO[0001] using existing policy` `                         ARN="arn:aws:iam::900736139882:policy/mattermost-apps-invoke-policy"` which makes sense put the tool is aware that I am now using a different bucket. @jfrerich helped me look into this and noticed that in this case the ` "Resource":` section of the policy does not update to include the manifest and static folders for the bucket set when the command is run. 

For cases where someone maybe using multiple staging/test buckets the script should be able to accommodate adding the S3 resources to the policy.

On a minor note, there is a debug command for `add-manifest` that serves no purpose in On-Prem mode. Manifests can be stored to the KV but there is no way to use them for installation after as only the 1 step install from URL is functional.

File: upstream/upaws/aws.go
Comment (stylianosrigas) [inline upstream/upaws/aws.go:9]: @levb will this be used for APPs invoke?

File: upstream/upaws/aws.go
Comment (stylianosrigas) [inline upstream/upaws/aws.go:48]: @levb Have you tested this? You may need to add this section too
```json
{
            "Effect": "Allow",
            "Action": [
                "s3:ListBucket",
                "s3:ListBucketVersions",
                "s3:GetBucketLocation"
            ],
            "Resource": [
                ""arn:aws:s3:::{{.Bucket}}"
            ]
        }
```

File: server/proxy/install.go
Comment (hanzei) [inline server/proxy/install.go:104]: My typo
```suggestion
		message = md.MD(fmt.Sprintf("Successfully installed %s", app.AppID))
```

File: upstream/upaws/clean.go
Comment (levb) [reply upstream/upaws/clean.go]: What's the difference?

**Comment by levb** [comment]
> For cases where someone maybe using multiple staging/test buckets the script should be able to accommodate adding the S3 resources to the policy.

@DHaussermann once the policy is initialized, they can manually edit it and add more bucket permissions.

File: upstream/upaws/aws.go
Comment (levb) [reply upstream/upaws/aws.go:9]: Yes

File: upstream/upaws/aws.go
Comment (levb) [reply upstream/upaws/aws.go:48]: Yes, we did (with **s, still a final round to be done). From https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html

You need the s3:GetObject permission for this operation. For more information, see Specifying Permissions in a Policy. If the object you request does not exist, the error Amazon S3 returns depends on whether you also have the s3:ListBucket permission.

If you have the s3:ListBucket permission on the bucket, Amazon S3 will return an HTTP status code 404 ("no such key") error.

If you don’t have the s3:ListBucket permission, Amazon S3 will return an HTTP status code 403 ("access denied") error.

Comment (levb) [reply upstream/upaws/aws.go:48]: Getting a 404 is fine

**Comment by DHaussermann** [comment]
> Once the policy is initialized, they can manually edit it and add more bucket permissions.

Thanks Lev, giving the low likelihood that this is needed - This sounds very reasonable to me. I'll give this a try for the next test pass now 👍

File: cmd/appsctl/main.go
Comment (levb) [reply cmd/appsctl/main.go:44]: Fixed, was a typo, and a very unfortunate one. Thx @hanzei for catching!

**Comment by DHaussermann** [comment]
@hanzei I realize that after the merge of your work, the `apps list` command now shows you the records in the manifest when running in developer mode for on-prem. 
I have a concern that this can cause confusion when troubleshooting. For instances in the cloud the apps are only listed when the is a record in the config **and** a corresponding lambda is available so only installable apps are listed here. Unless we can also indicate in the table which apps are installable, this change may cause confusion. Any thoughts?

**Comment by hanzei** [comment]
@DHaussermann For Cloud, all apps listed there are installable.

For on-prem, the lambda might get deleted and then an app can't be installed any longer. Given that this issue can also also happen while the app is running it seems like a different issue to me. Did I got your question correctly?

**Comment by DHaussermann** [comment]
@hanzei yes, I believe if you understand the concern. My question would be, is it not preferable to keep the same logic for on-prem and have only installable apps listed in the table rather than showing whatever is in the plugin manifest?

File: server/proxy/install.go
Comment (levb) [reply server/proxy/install.go:96]: Unfortunately, we'll need a little refactoring to properly differentiate "framework-introduced errors" such as 404 from the App-introduced errors. In the case on Install, 404s should be ignorable but they are not. See https://mattermost.atlassian.net/browse/MM-35152

**Comment by hanzei** [comment]
@DHaussermann What is the difference between an "installable app" and one with a registered manifest?

File: server/plugin.go
Comment (hanzei) [inline server/plugin.go:101]: Actually, my initial comment was invalid. The marketplace install is only available in Cloud. Given that the store most likely won't work for devs, I'm inclined revert this change.
```suggestion
	if conf.MattermostCloudMode {
```

**Comment by DHaussermann** [comment]
@hanzei unless I'm mistaken, for the cloud deployments - Apps are only listed in the table if there is also a lambda function with the corresponding name available. Therefore, apps are only listed in the table if they are available to install.

**Comment by hanzei** [comment]
@DHaussermann That is not the case. Both for cloud and on-prem there can be apps listed that don't have a corresponding lambda function.

**Comment by hanzei** [comment]
@DHaussermann I don't get which case you are describing. The lambda should work unless it get manually deleted, which IMO we should not deal with right now.

**Comment by DHaussermann** [comment]
@hanzei sorry it seems maybe my concern above was not clear or I am making an incorrect assumption. 
1. My understanding is that with your changes merged in, Apps will always show in the table if they are listed in the manifest regardless of if there is also a lambda available.
2. This differs from Cloud installs where the record will show in the table if there is a record in the config **AND** the lambda is available.

Are these 2 points accurate?

**Comment by hanzei** [comment]
1. Technical correct, but this should not happen. Apps running in AWS are added by `/apps install aws`. It fetches the manifest from S3 which is only present if `appsctl` was run successfully before. When should this case happen?
2. Technical not correct. If we miss configure things, the lambda might not be available.

**Comment by hanzei** [comment]
@DHaussermann Is this conversation a blocker to merging the PR?

File: server/plugin.go
Comment (hanzei) [reply server/plugin.go:101]: Applying this changes as the current behavior also confused @DHaussermann

**Comment by levb** [comment]
👍

File: server/command/install.go
Comment (catalintomai) [inline server/command/install.go:16]: maybe add a comment for future readers that for this path (Marketplace) we've previously added the manifests to the store (using `InitGlobal`?).

File: server/plugin.go
Comment (catalintomai) [inline server/plugin.go:81]: what does it actually mean (practically) to be in both `MattermostCloudMode` and `DeveloperMode`?

File: server/config/service.go
Comment (catalintomai) [inline server/config/service.go:90]: we can get nil the second time (and more) as well, no? Is the on-prem mode automatically the non-MattermostCloud mode?

Comment (levb) [reply server/config/service.go:90]: Yes, and at this point we assume "Self-managed", see https://github.com/mattermost/mattermost-plugin-apps/blob/181555e6ed4263c6e94a8a4e26706a1bc6192010/server/config/config.go#L124

Comment (hanzei) [reply server/config/service.go:90]: Should we log an error in that case?

File: server/plugin.go
Comment (hanzei) [reply server/plugin.go:81]: They are not mutually exclusive.

If an server is in `MattermostCloudMode` and in `DeveloperMode` it means that:
- There will be debug command like `/apps debug-list-bindings`
- `manifests.json` will be used to fetch manifest from S3
- There will only be a `/apps install marketplace` command

File: server/config/service.go
Comment (hanzei) [inline server/config/service.go:93]:
```suggestion
		license = s.mm.System.GetLicense()
		if license == nil {
			s.mm.Log.Warn("Failed to fetch license two times. Falling back to on-prem mode.")
		}
```

File: server/config/service.go
Comment (hanzei) [reply server/config/service.go:90]: Suggestion: https://github.com/mattermost/mattermost-plugin-apps/pull/191#discussion_r647406090

**Comment by DHaussermann** [comment]
Testing for this has completed. 
No further changes are needed yo on-prem mode.

I did find 1 point of concern when regression testing cloud mode. Install command - including for cloud - have been updated. I was not aware that clous install had also changed but, cloud install commands now look like `/apps install marketplace com.mattermost.servicenow`

@hanzei  this changes breaks the marketplace installation when the plugin is running in Cloud mode as the "_Install_" button is running the command without the installation type of `marketplace`

This is concerning to me because it means we need webapp changes to accommodate and will once again have a dependency on when can deploy based on when the webapp change is on Cloud. If this is the way we must implement the cloud installation command it does not block this PR. But we need a corresponding webapp PR and a fixed deployment date.   
cc @levb and @catalintomai

**Comment by hanzei** [comment]
@DHaussermann We can add a temporary workaround so the old command still works. Let me do that in a separate PR if you don't mind.

**Comment by DHaussermann** [comment]
@hanzei to clarify ☝️ you mean cloud mode will not require the installation type parameter (or fail-over to `marketplace` when not supplied)? 

This makes sense to me. We can then co-ordinate this change separately.

**Comment by levb** [comment]
How about we merge this PR for the on-prem, and if we still need to figure out the command syntax for 0.7 we do a separate PR/ticket for it?

**Comment by hanzei** [comment]
@DHaussermann Exactly.

@levb Sounds good to me.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Tested deployment to my personal AWS instance
- Tested that multiple apps can be deployed and are functional
- Tested use of appctl tool as documented
- Worked with Dev on corresponding doc PR as an instruction set
- Confirmed the app runs in cloud mode when legacy AWS keys are present
- Regression tested Cloud mode
- As per @hanzei and @levb we will be decided and addressed in a separate PR before 0.7.0 release. Created https://mattermost.atlassian.net/browse/MM-36294 to address this.

LGTM!
Please merge.

**Comment by hanzei** [comment]
Thanks for the testing @DHaussermann :+1: 

I'm going to merge this PR as it's a great improvement in it self and blocked some other PR right now due to code dependencies and merge conflicts.

@stylianosrigas @catalintomai Please leave any outstanding comments in this PR and I will resolve them in a separate one.

**Pull request merged**
